### PR TITLE
Adopt affix/overlay service

### PR DESCRIPTION
**Pull request by Blackbaud-SteveBrush**
Issue(s):
https://github.com/blackbaud/skyux-popovers/issues/43
https://github.com/blackbaud/skyux-lookup/issues/39
https://github.com/blackbaud/skyux-ag-grid/issues/19

NOTE: I made a few accessibility changes to align the Dropdown with the W3C guidelines: https://www.w3.org/WAI/tutorials/menus/application-menus/#keyboard-behavior

### CODE REVIEW

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2](https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2)

File: src/app/visual/popover/popover-visual.component.html [hunk 1]
```diff
@@ -1,51 +1,16 @@
...
-  <div class="popover-visual-container">
-    <sky-popover placement="left" popoverTitle="Did you know?">
-      The content of a popover can be text, HTML, or Angular components.
-    </sky-popover>
-  </div>
+<div
+  class="static-popovers"
+  id="screenshot-all-popovers"
+>
+  <ng-container
```
Comment (Blackbaud-SteveBrush) [inline src/app/visual/popover/popover-visual.component.html]: Because popovers are now generated dynamically in an overlay, we need to print them out differently.

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2](https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2)

File: src/app/public/modules/dropdown/dropdown-menu.component.ts [hunk 2]
```diff
@@ -284,4 +282,130 @@ export class SkyDropdownMenuComponent implements AfterContentInit, OnDestroy {
...
+        /* istanbul ignore else */
+        if (selectedItem) {
+          this.menuChanges.next({
+            selectedItem
+          });
+        }
+      });
+
+    Observable
+      .fromEvent(dropdownMenuElement, 'keyup')
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/dropdown/dropdown-menu.component.ts]: I noticed that when pressing 'enter' on any menu items, it will activate that item, but then move focus to the top of the DOM. With the current version, pressing `enter` will still keep focus on the dropdown. Should we keep this behavior?

File: src/app/public/modules/dropdown/dropdown-menu.component.ts [hunk 3]
```diff
@@ -284,4 +282,130 @@ export class SkyDropdownMenuComponent implements AfterContentInit, OnDestroy {
...
+      .takeUntil(this.ngUnsubscribe)
+      .subscribe((event: KeyboardEvent) => {
+        const key = event.key.toLowerCase();
+
+        /*tslint:disable-next-line:switch-default*/
+        switch (key) {
+          case 'escape':
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown-menu.component.ts]: The ideal setup is `keyup` because it reflects a deliberate action from the user (i.e., `keydown` would fire if you pressed a key down but didn't remove your finger from the key). However, for some keys (like arrow keys and tab), JavaScript performs native functionality on the `keydown` action, which we need to short-circuit to perform our own actions. (These events should be copied from the original implementation, for the most part.)

File: package.json [hunk 4]
```diff
@@ -43,43 +43,40 @@
...
-    "@blackbaud/skyux-lib-stache": "3.2.3",
-    "@pact-foundation/pact": "9.6.0",
-    "@pact-foundation/pact-web": "9.6.0",
-    "@skyux-sdk/builder": "3.13.0",
-    "@skyux-sdk/builder-plugin-skyux": "1.2.1",
+    "@blackbaud/skyux-lib-stache": "3.3.0",
+    "@skyux-sdk/builder": "3.16.2",
+    "@skyux-sdk/builder-plugin-skyux": "1.3.1",
     "@skyux-sdk/e2e": "3.1.2",
-    "@skyux-sdk/pact": "3.3.1",
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply package.json]: Yeah, the pact stuff isn't needed for our components, and it will be removed in the next version of SKY UX, so I thought we could go ahead and remove it now. It shouldn't have any effect on our consumers or the build result.

File: src/app/public/modules/dropdown/types/index.ts [hunk 5]
```diff
@@ -1,3 +1,4 @@
+export * from './dropdown-horizontal-alignment';
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/types/index.ts]: I've done this already in the `rc-4.0.0` branch, so when we migrate this branch to RC, it won't be much of a problem.

File: src/app/public/modules/dropdown/dropdown-extensions.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import {
+  SkyAffixHorizontalAlignment
+} from '@skyux/core';
+
+import {
+  SkyDropdownHorizontalAlignment
+} from './types/dropdown-horizontal-alignment';
+
+export function parseAffixHorizontalAlignment(
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown-extensions.ts]: I want to check that specific string values are represented by the dropdown's specific type. It should throw an error if a new type is created by the affix module, but not represented by the dropdown so that we can make the necessary changes without affecting the consumer.

[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown-extensions.ts]: I had to do something similar for the C# notifications service and thought it would apply similarly here.

File: src/app/public/modules/dropdown/dropdown-menu.component.html [hunk 7]
```diff
@@ -1,5 +1,5 @@
 <div
-  class="sky-dropdown-menu"
+  class="sky-dropdown-menu sky-shadow sky-rounded-corners"
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown-menu.component.html]: Good catch. I'll make an adjustment.

File: src/app/public/modules/dropdown/dropdown.component.spec.ts [hunk 8]
```diff
@@ -1,816 +1,1130 @@
...
 
-  function getDropdownMenuHostElement(): HTMLElement {
-    return fixture.nativeElement.querySelector('sky-dropdown-menu') as HTMLElement;
+  function detectChangesFakeAsync(): void {
+    fixture.detectChanges();
+    tick();
+    fixture.detectChanges();
+    tick();
+    fixture.detectChanges();
+    tick();
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown.component.spec.ts]: There are a few setTimeouts and observable streams we have to accommodate. I suppose I can add a comment.

File: src/app/public/modules/dropdown/dropdown.component.html [hunk 9]
```diff
@@ -57,21 +52,20 @@
...
-    (popoverOpened)="onPopoverOpened()"
-    #popover
+</div>
+
+<ng-template
+  #menuContainerTemplateRef
+>
+  <div
+    class="sky-dropdown-menu-container"
+    [class.hidden]="!isVisible"
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown.component.html]: Good idea.

File: src/app/public/modules/dropdown/dropdown.component.spec.ts [hunk 10]
```diff
@@ -1,816 +1,1130 @@
...
+      container = getMenuContainerElement();
+
+      // Tab key should progress to next item after the trigger button.
+      expect(container).toBeNull();
     }));
 
-    it('should allow focusing trigger button', fakeAsync(() => {
-      tick();
-      fixture.detectChanges();
+    xit('should ignore menu keyboard events if menu is closed', fakeAsync(() => {
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown.component.spec.ts]: Oops! Good catch.

File: src/app/public/modules/dropdown/types/dropdown-horizontal-alignment.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+import {
+  SkyAffixHorizontalAlignment
+} from '@skyux/core';
+
+export type SkyDropdownHorizontalAlignment = SkyAffixHorizontalAlignment;
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/types/dropdown-horizontal-alignment.ts]: The consumer should only know about Dropdown's types. We don't want to be in a place where the implementation details leak out from our Input/Outputs (see the `alignment` input as a bad example of this; that input is forever tied to popovers, even though we're not using popovers as the placement mechanism anymore: https://github.com/blackbaud/skyux-popovers/blob/5e166e2c6592562653894b666612536e82a422e0/src/app/public/modules/dropdown/dropdown.component.ts#L72).

File: src/app/public/modules/dropdown/dropdown.component.ts [hunk 12]
```diff
@@ -204,91 +291,170 @@ export class SkyDropdownComponent implements OnInit, OnDestroy {
...
-            this.isKeyboardActive = true;
+    Observable
+      .fromEvent(buttonElement, 'keyup')
+      .takeUntil(this.ngUnsubscribe)
+      .subscribe((event: KeyboardEvent) => {
+        const key = event.key.toLowerCase();
+        if (this.isOpen && key === 'escape') {
+          this.sendMessage(SkyDropdownMessageType.Close);
+          this.sendMessage(SkyDropdownMessageType.FocusTriggerButton);
+          event.stopPropagation();
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown.component.ts]: If you put a dropdown menu in a popover and the user presses the escape key, it would close both the dropdown and the popover in one keypress (unless we stop the propagation). I couldn't figure out a way to do it without stopping propagation unless you have an idea?

File: src/app/public/modules/dropdown/dropdown.component.ts [hunk 13]
```diff
@@ -303,19 +469,25 @@ export class SkyDropdownComponent implements OnInit, OnDestroy {
...
-      'below',
-      this.alignment
-    );
-  }
+  private positionDropdownMenu(): void {
+    this.isVisible = false;
+    this.createOverlay();
+    this.changeDetector.markForCheck();
+
+    this.windowRef.getWindow().setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown.component.ts]: This was kept in because it was in the previous implementation (trying to avoid needless changes), but if you don't mind, I'll remove it; I'm not a fan of using `windowRef` if it can be avoided.

File: src/app/public/modules/popover/popover.directive.ts [hunk 14]
```diff
@@ -132,80 +133,98 @@ export class SkyPopoverDirective implements OnChanges, OnDestroy {
...
+
     Observable
       .fromEvent(element, 'keyup')
-      .takeUntil(this.idled)
+      .takeUntil(this.ngUnsubscribe)
       .subscribe((event: KeyboardEvent) => {
         const key = event.key.toLowerCase();
-        if (key === 'escape' && this.isPopoverOpen()) {
-          event.stopPropagation();
-          event.preventDefault();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/popover/popover.directive.ts]: If we remove `stopPropagation`, what will happen if you hit `escape` on a popover that's inside of a modal? Will it close the modal now that the event bubbles?

File: src/app/public/modules/popover/popover.directive.spec.ts [hunk 15]
```diff
@@ -1,309 +1,850 @@
...
-      expect(positionNextToSpy).not.toHaveBeenCalled();
-    }
+describe('Popover directive', () => {
 
-    SkyAppTestUtility.fireDomEvent(caller.nativeElement, openTrigger);
-    expect(positionNextToSpy).toHaveBeenCalled();
+  let fixture: ComponentFixture<PopoverFixtureComponent>;
 
-    callerInstance.skyPopover.isOpen = true;
+  function getCallerElement(): HTMLButtonElement {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/popover/popover.directive.spec.ts]: Would you mind wrapping these helpers in a collapsable region?

File: src/app/public/modules/popover/popover.directive.ts [hunk 16]
```diff
@@ -132,80 +133,98 @@ export class SkyPopoverDirective implements OnChanges, OnDestroy {
...
+        switch (key) {
+          case 'arrowdown':
+          case 'arrowleft':
+          case 'arrowright':
+          case 'arrowup':
+          case 'down':
+          case 'left':
+          case 'right':
+          case 'up':
+            this.sendMessage(SkyPopoverMessageType.Focus);
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/popover/popover.directive.ts]: In the current popover, the arrow keys will NOT focus on the popover. Is this change deliberate?

File: src/app/public/modules/popover/popover.component.html [hunk 17]
```diff
@@ -1,52 +1,5 @@
-<div
-  class="sky-popover-container"
-  tabindex="-1"
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/popover/popover.component.html]: The current popover is not focusable. It looks like this change may have made them focusable - was that deliberate?

File: src/app/public/modules/dropdown/dropdown-menu.component.ts (hunk 2 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown-menu.component.ts]: Good catch!

File: src/app/visual/popover/popover-visual.component.html [hunk 18]
```diff
@@ -310,12 +292,12 @@
...
+      Above, right
     </button>
   </p>
 </div>
 <div id="screenshot-popover-fullscreen">
-    <button
+  <button
     id="btn-popover-fullscreen"
     class="sky-btn"
     [skyPopover]="popoverFullscreen"
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/popover/popover-visual.component.html]: When I click on the "Show fullscreen popover" button, and then hit `tab`, it closes the popover. Shouldn't it set focus inside of the popover so I can iterate over the tabbable elements?

File: src/app/public/modules/popover/popover-content.component.html [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+  [@.disabled]="!enableAnimations"
+  [@skyPopoverAnimation]="animationState"
+  [class.sky-popover-hidden]="!placement"
+  [class.sky-popover-max-height]="!allowFullscreen"
+  [ngClass]="[
+    'sky-popover-alignment-' + horizontalAlignment,
+    'sky-popover-placement-' + placement
+  ]"
+  (@skyPopoverAnimation.start)="onAnimationEvent($event)"
+  (@skyPopoverAnimation.done)="onAnimationEvent($event)"
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover-content.component.html]: LOL, they were originally, but I thought you were going to ask them to be ordered alphabetically (inputs then outputs). Let me know if you want me to revert it.

File: src/app/public/modules/popover/popover-content.component.ts [hunk 20]
```diff
@@ -0,0 +1,377 @@
...
+
+  public popoverTitle: string;
+
+  @ViewChild('arrowRef', { read: ElementRef })
+  private arrowRef: ElementRef;
+
+  @ViewChild('popoverRef', { read: ElementRef })
+  private popoverRef: ElementRef;
+
+  @ViewChild('contentTarget', { read: ViewContainerRef })
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover-content.component.ts]: I don't think we need to at this time since we don't know for sure if these should be static or not.

File: src/app/public/modules/popover/popover-content.component.ts [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+    this.contentTarget.createEmbeddedView(this.context.contentTemplateRef);
+    this.addEventListeners();
+  }
+
+  public ngOnDestroy(): void {
+    this.ngUnsubscribe.next();
+    this.ngUnsubscribe.complete();
+
+    this._closed.complete();
+    this._opened.complete();
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover-content.component.ts]: Good catch!

File: src/app/public/modules/popover/popover-content.component.ts [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+      return;
+    }
+
+    if (this.placement === 'fullscreen' && this.allowFullscreen) {
+      this.isOpen = true;
+      this.changeDetector.markForCheck();
+      return;
+    }
+
+    setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover-content.component.ts]: Good call.

File: src/app/public/modules/popover/popover.directive.ts (hunk 14 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover.directive.ts]: Hm, you're right. I've re-added these.

File: src/app/public/modules/popover/popover.directive.spec.ts (hunk 15 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover.directive.spec.ts]: You got it.

File: src/app/public/modules/popover/popover.directive.ts (hunk 16 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover.directive.ts]: Yes. Today, when the user presses the "tab" key, the popover content receives focus (since it exists in the DOM directly after the button). However, according to W3C the tab key should not bring focus to a "modal" or tooltip component (instead, the modal should immediately receive focus when it is opened; however, this is not an expected behavior of popovers today). I added arrow key focusability so that the user could use the keyboard to navigate to the popover (which is now in an overlay). I checked our issue history and it turns out that popovers never received an a11y audit from Matt Gregg. It might be worth reevaluating the popover's interactions to meet W3C guidelines.

File: src/app/public/modules/popover/popover.component.html (hunk 17 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/popover/popover.component.html]: See above.

File: src/app/visual/popover/popover-visual.component.html (hunk 18 above)
Comment (Blackbaud-SteveBrush) [reply src/app/visual/popover/popover-visual.component.html]: Ugh, you're right. This feature is going to die in the next version of SKY UX (and our design guidelines say to NOT use fullscreen popovers). Let me see if I can come up with something simple to address this.

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2](https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2)

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2](https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2)

File: src/app/public/modules/dropdown/types/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/dropdown/types/index.ts:1]: Forgot about the 4.0 branch! 👍

File: src/app/public/modules/dropdown/dropdown-extensions.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/dropdown/dropdown-extensions.ts:9]: Seems prudent enough. 👍

File: src/app/public/modules/dropdown/types/dropdown-horizontal-alignment.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/dropdown/types/dropdown-horizontal-alignment.ts:5]: Ah, yes. That makes perfect sense. Thanks for the explanation!

File: src/app/public/modules/dropdown/dropdown.component.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/dropdown/dropdown.component.ts]: Me too. Yeah, go for it.

File: src/app/public/modules/popover/popover-content.component.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/popover/popover-content.component.ts:121]: Not a deal breaker. Just wanted to be aware it might need a change in the future.

File: src/app/public/modules/dropdown/dropdown.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/dropdown/dropdown.component.ts]: Yeah, looking at the rest of my review, I think I also realized this challenge when looking at your `SkyPopoverDirective` changes. It's necessary there too. Since this doesn't take into account any click targets or focus, I feel like the only way to do this is with the `isOpen` check and then stop propagation. Let's keep it.

File: src/app/visual/popover/popover-visual.component.html (hunk 18 above)
Comment (Blackbaud-AlexKingman) [reply src/app/visual/popover/popover-visual.component.html:303]: Nice fix.

File: src/app/public/modules/popover/popover-content.component.html (hunk 19 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/popover/popover-content.component.html:11]: What you've got is fine. Honestly, I don't know our rule on that. Maybe we should clarify in the docs a little better 😆 Let's move on!

File: src/app/public/modules/popover/popover.directive.ts (hunk 16 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/popover/popover.directive.ts:160]: Sounds good. Thanks for looking into this. I can confirm the arrow keys work as you described. 👍

File: src/app/public/modules/dropdown/dropdown-menu.component.ts [hunk 23]
```diff
@@ -284,4 +282,130 @@ export class SkyDropdownMenuComponent implements AfterContentInit, OnDestroy {
...
+                this.sendMessage(SkyDropdownMessageType.Close);
+                this.sendMessage(SkyDropdownMessageType.FocusTriggerButton);
+              } else {
+                this.sendMessage(SkyDropdownMessageType.FocusPreviousItem);
+              }
+            } else {
+              const isLastItem = (focusableItems[focusableItems.length - 1] === event.target);
+              if (isLastItem) {
+                this.sendMessage(SkyDropdownMessageType.Close);
+                this.sendMessage(SkyDropdownMessageType.FocusTriggerButton);
```
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/dropdown/dropdown-menu.component.ts]: Looks like this is the same now. Thanks!

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2](https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2)

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2](https://developer.blackbaud.com/skyux-popovers?_sp.spa=pr-adopt-overlay-service-2)

File: src/app/public/modules/dropdown/dropdown-menu.component.ts [hunk 24]
```diff
@@ -284,4 +282,130 @@ export class SkyDropdownMenuComponent implements AfterContentInit, OnDestroy {
...
+          // search bar.
+          case 'tab':
+            const focusableItems = this.coreAdapterService.getFocusableChildren(
+              dropdownMenuElement
+            );
+
+            if (event.shiftKey) {
+              const isFirstItem = (focusableItems[0] === event.target);
+              if (isFirstItem) {
+                this.sendMessage(SkyDropdownMessageType.Close);
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/dropdown/dropdown-menu.component.ts]: This behavior should be adjusted based on the updated accessibility features. Let me know if it doesn't work as expected.
